File: src/driver_bot_cpp/nodes/src/Perceptions/distance_node.cpp
//#include <driver_bot_cpp/distance_node.h>
#include "distance_node.h"

DistanceDetection::DistanceDetection(const float angle1, const float angle2)
:m_pubDistance{m_node.advertise<std_msgs::Float32>("distance", 10)},
 m_pubMinDistance{m_node.advertise<std_msgs::Float32>("minDistance", 10)},
 m_subLidar{m_node.subscribe("rplidar", 10, &DistanceDetection::LidarCallback, this)},
 m_operationsDistance{angle1, angle2}
{
        
}

//Main functionalities
void DistanceDetection::LidarCallback(const driver_bot_cpp::lidar::ConstPtr& msg)
{
    /*Callback function containing data from rplidar
     *   Args:
     *       msg: contains data for distance and angle_increment
     */
    m_operationsDistance.SetDist(msg -> distance);
    m_operationsDistance.SetAngleIncrement(msg -> angle_increment);
}

void DistanceDetection::Run()
{
    /*Computes distance in range of 11/6 to 1/6 pi and minimum distance. Hereafter,
    the values are published on "distance" and "minDistance" topics.
    */
    float distanceRange;
    float minDistance;

    ros::Rate rate(10);

    while (ros::ok())
    {
        std::vector<float> angleArray = m_operationsDistance.DataAnalyses();
        distanceRange = m_operationsDistance.Detection(angleArray);
        std::vector<float> distance = m_operationsDistance.GetDist(); //can I prevent this copy here?
        minDistance = m_operationsDistance.MinimumDistance(distance);

        PublishDistanceRange(distanceRange);
        PublishMinDistance(minDistance);

        ros::spinOnce();
        rate.sleep();
    }
}

//Additional functionalities
void DistanceDetection::PublishDistanceRange(const float distance)
{
    /*Publishes distance for a specific angle range*/
    std_msgs::Float32 msg;
    msg.data = distance;
    m_pubDistance.publish(msg);
}

void DistanceDetection::PublishMinDistance(const float distance)
{
    /*Publishes minimum distance for 0 to 2 pi*/
    std_msgs::Float32 msg;
    msg.data = distance;
    m_pubMinDistance.publish(msg);
}


int main(int argc, char **argv)
{
    ros::init(argc, argv, "distanceDetection");
    float angle1 = ANGLE_ONE_DOMAIN_AT_FRONT_OF_ROBOT * M_PI, angle2 = ANGLE_TWO_DOMAIN_AT_FRONT_OF_ROBOT * M_PI;
    DistanceDetection dist{angle1, angle2};
    dist.Run();
    return 0;
}

File: catkin_work/src/driver_bot_pkg/src/python/Perceptions/distance_node.py
#!/usr/bin/env python3

import cv2
import rospy
import numpy as np
from std_msgs.msg import Bool
from std_msgs.msg import Float64
from driver_bot_pkg.msg import LidarData

# own packages
import Sensors.lidar_node as lidar_node
import Perceptions.PerceptionInterface as PerceptionInterface


class DistanceDetection():
    def __init__(self):

        self.pubDistance = rospy.Publisher('distance', Float64, queue_size=10)
        self.pubMinDistance = rospy.Publisher(
            'minDistance', Float64, queue_size=10)
        self.subLidar = rospy.Subscriber(
            'rplidar', LidarData, self.lidarCallback)

        self.angle_increment = 1
        self.dist = 10

    def lidarCallback(self, msg):
        """Callback function containing data from rplidar
        Args:
            msg: contains data for distance and angle_increment
        """
        self.angle_increment = msg.angle_increment
        self.dist = msg.distance

    def dataAnalyses(self):
        """Produces an array of of values with increment self.angle_increment for one round
        trip between 0 and 2 pi.We reverse the list since the rpLidar scanner starts at 2pi
        and ends at 0 pi"""

        numberArray = np.linspace(
            0, 2, 2 * np.pi / self.angle_increment)  # idx of angles
        return np.flip(numberArray, 0)

    def detection(self):
        """Breaks up distance object from 0 to 2pi into a specific angle range
        Variables:
            angle1: lower bound of angle range
            angle2: upper bound of angle range

        Returns:
            array of distance values for angle1 < angle < angle2
        """
        angle1 = 1/6
        angle2 = 11/6

        angleArray = self.dataAnalyses()
        distanceIdx = np.where((angleArray > angle1) & (angleArray < angle2))[
            0]  # get unwanted elements

        indicesArray = np.arange(len(angleArray))
        # delete unwanted elements

        angleIdx = np.delete(indicesArray, distanceIdx)
        try:
            distanceRange = self.getIndices(angleIdx, self.dist)
        except:
            distanceRange = [1]

        return self.minimumDistance(distanceRange)

    def run(self):
        rate = rospy.Rate(10)

        while not rospy.is_shutdown():

            distanceRange = self.detection()
            minDistance = self.minimumDistance(self.dist)

            self.publishDistanceRange(distanceRange)
            self.publishMinDistance(minDistance)

            rate.sleep()

    # ---------Additional functionalities----------#
    def publishDistanceRange(self, distance):
        """Publishes distance for a specific angle range"""
        self.pubDistance.publish(distance)
        print('dist range: ' + str(distance))

    def publishMinDistance(self, distance):
        """Publishes minimum distance for 0 to 2 pi"""
        self.pubMinDistance.publish(distance)
        print('min dist is: ' + str(distance))

    # ---------support functions----------#

    def minimumDistance(self, distance):
        """Function calculates minimum distance to object
        Args:
            distance: array containing distance to all objects between angle1 and angle2 for 0.15m < d < 12m

        Returns:
            minimum distance of the provided distance array
        """
        distMin = np.min(distance)
        return distMin

    def getIndices(self, arrayIdx, arrayObj):
        """Retrieves values from arrayObj based on indices provided in arrayIdx
        Args:
            arrayidx: array containing wanted indices

            arrayObj: array containing objects
        Returns: Array with wanted values, based on elements provided in arrayIdx
        """
        obj = np.zeros(len(arrayIdx) - 1)
        arrayIdx = arrayIdx[0: len(arrayIdx) - 1]

        for idx, value in enumerate(arrayIdx):
            obj[idx] = arrayObj[value]

        return obj


if __name__ == '__main__':
    rospy.init_node('distanceDetection')
    dist = DistanceDetection()
    dist.run()
